source/base/qprojector.cc

```diff
@@ -884,40 +867,18 @@ QProjector<3>::project_to_all_faces(const ReferenceCell      &reference_cell,
                 Point<3> mapped_point;
 
                 // map reference quadrature point
-                for (unsigned int i = 0; i < n_linear_shape_functions; ++i)
-                  mapped_point +=
-                    support_points[i] *
-                    poly.compute_value(i, sub_quadrature_points[j]);
+                for (const unsigned int i :
+                     face_reference_cell.vertex_indices())
+                  mapped_point += support_points[i] *
+                                  face_reference_cell.d_linear_shape_function(
+                                    sub_quadrature_points[j], i);
 
                 points.push_back(mapped_point);
 
-                // scale quadrature weight
-                const double scaling = [&]() {
-                  const unsigned int dim_     = 2;
-                  const unsigned int spacedim = 3;
-
-                  Tensor<1, dim_, Tensor<1, spacedim>> DX_t;
-
-                  shape_derivatives.resize(n_linear_shape_functions);
-
-                  for (unsigned int i = 0; i < n_linear_shape_functions; ++i)
-                    shape_derivatives[i] =
-                      poly.compute_1st_derivative(i, sub_quadrature_points[j]);
-
-                  for (unsigned int k = 0; k < n_linear_shape_functions; ++k)
-                    for (unsigned int i = 0; i < spacedim; ++i)
-                      for (unsigned int j = 0; j < dim_; ++j)
-                        DX_t[j][i] +=
-                          shape_derivatives[k][j] * support_points[k][i];
-
-                  Tensor<2, dim_> G;
-                  for (unsigned int i = 0; i < dim_; ++i)
-                    for (unsigned int j = 0; j < dim_; ++j)
-                      G[i][j] = DX_t[i] * DX_t[j];
-
-                  return std::sqrt(determinant(G));
-                }();
-
+                // rescale quadrature weights so that the sum of the weights on
+                // each face equals the measure of that face.
+                const double scaling = reference_cell.face_measure(face_no) /
+                                       face_reference_cell.volume();
```

Comment:
This has gotten pleasingly easier! Did we previously miss that the determinant is constant throughout the face and could be pre-computed? What an oversight :-)

Reply:
We took a lot of shortcuts while writing the simplex code. We also didn't have `ReferenceCell::face_measure()` at this point.

Reply:
I appreciate your clean-up efforts!